packages/commonwealth/test/devnet/cosmos/tokenBalanceFetching.spec.ts

```diff
@@ -420,4 +447,167 @@ describe('Token Balance Cache Cosmos Tests', function () {
+        balanceSourceType: BalanceSourceType.CW20,
+        addresses: [discobotAddress],
+        sourceOptions: {
+          cosmosChainId,
+          contractAddress,
+        },
+        cacheRefresh: true,
+      });
+
+      expect(Object.keys(balance).length).to.equal(0);
+    });
+    it('should not throw if a single address out of many fails', async () => {
+      const balance = await tokenBalanceCache.getBalances({
+        balanceSourceType: BalanceSourceType.CW20,
+        addresses: [addressWithToken, badAddress],
+        sourceOptions: {
+          cosmosChainId,
+          contractAddress,
+        },
+        cacheRefresh: true,
+      });
+
+      expect(Object.keys(balance).length).to.equal(1);
+      expect(balance[addressWithToken]).to.equal('3000000');
+    });
+    it('should correctly batch balance requests', async () => {
+      const bulkAddresses = await generateCosmosAddresses(20);
+      bulkAddresses.splice(4, 0, addressWithToken);
+      bulkAddresses.splice(5, 0, addressWithoutToken);
+      const balances = await tokenBalanceCache.getBalances({
+        balanceSourceType: BalanceSourceType.CW20,
+        addresses: bulkAddresses,
+        sourceOptions: {
+          cosmosChainId,
+          contractAddress,
+        },
+        cacheRefresh: true,
+        batchSize: 5,
+      });
+      expect(Object.keys(balances).length).to.equal(22);
+      expect(balances[addressWithToken]).to.equal('3000000');
+      expect(balances[addressWithoutToken]).to.equal('0');
+    });
+
+    describe('Caching CW20', () => {
+      const balanceTTL = 20;
+
+      before('Set TBC caching TTL and reset Redis', async () => {
+        await cache().flushAll();
+      });
+
+      it('should cache balance, then refresh after TTL', async () => {
+        const chainNode = await models.ChainNode.findOne({
+          where: {
+            cosmos_chain_id: cosmosChainId,
+          },
+        });
+        const tmClient = await tokenBalanceCache.getTendermintClient({
+          chainNode,
+        });
+        const api = QueryClient.withExtensions(tmClient, setupWasmExtension);
+
+        const key = {
+          balance: {
+            address: addressWithToken,
+          },
+        };
+        const response = await api.wasm.queryContractSmart(
+          contractAddress,
+          key,
+        );
+        const expectedAddressOneBalance = response.balance.toString();
+
+        const balance = await tokenBalanceCache.getBalances(
+          {
+            balanceSourceType: BalanceSourceType.CW20,
+            addresses: [addressWithToken],
+            sourceOptions: {
+              cosmosChainId,
+              contractAddress,
+            },
+            cacheRefresh: true,
+          },
+          balanceTTL,
+        );
+
+        expect(Object.keys(balance).length).to.equal(1);
+        expect(balance[addressWithToken]).to.equal(expectedAddressOneBalance);
+
+        delay(20000);
```

Comment:
Does the delay need to be this long?

Reply:
Yeah, this is copied from previous cache-related tests. In these cases, the TTL is 20 seconds, so we are deliberately testing against that here.

Reply:
Should have had await here though. Added now.